Python 3.12: `PyUnicode_AS_UNICODE` has been removed

##### Description:
In Python 3.12, `PyUnicdeo_AS_UNICODE` has been removed:

https://peps.python.org/pep-0623/#python-3-12

It's still used in https://github.com/urwid/urwid/blob/f7d2f2f6a8a519bd64ebbe1895e5c6b5948199f3/source/str_util.c

##### Affected versions (if applicable)
- [x] `master` branch (specify commit)
- [x] Latest stable version from `pypi`
- [ ] Other (specify source)

##### Steps to reproduce (if applicable)
Attempting to build urwid will give errors of this form:

```
gcc ... -fPIC -I/usr/include/python3.12 -c source/str_util.c -o
build/temp.linux-x86_64-cpython-312/source/str_util.o
source/str_util.c: In function ‘Py_IsWideChar’:
source/str_util.c:484:16: warning: implicit declaration of function
‘PyUnicode_AS_UNICODE’; did you mean ‘PyUnicode_AsUCS4’?
[-Wimplicit-function-declaration]
  484 |         ustr = PyUnicode_AS_UNICODE(text);
      |                ^~~~~~~~~~~~~~~~~~~~
      |                PyUnicode_AsUCS4
source/str_util.c:484:14: warning: assignment to ‘Py_UNICODE *’ {aka
‘int *’} from ‘int’ makes pointer from integer without a cast
[-Wint-conversion]
  484 |         ustr = PyUnicode_AS_UNICODE(text);
      |              ^
source/str_util.c: In function ‘Py_CalcWidth’:
source/str_util.c:663:14: warning: assignment to ‘Py_UNICODE *’ {aka
‘int *’} from ‘int’ makes pointer from integer without a cast
[-Wint-conversion]
  663 |         ustr = PyUnicode_AS_UNICODE(text);
      |              ^
source/str_util.c: In function ‘Py_CalcTextPos’:
source/str_util.c:749:14: warning: assignment to ‘Py_UNICODE *’ {aka
‘int *’} from ‘int’ makes pointer from integer without a cast
[-Wint-conversion]
  749 |         ustr = PyUnicode_AS_UNICODE(text);
      |              ^

```

##### Expected/actual outcome
[...]

--- a/.github/workflows/pythonpackage.yml
+++ b/.github/workflows/pythonpackage.yml
@@ -13,7 +13,7 @@ jobs:
       max-parallel: 6
       matrix:
         os: [ "ubuntu-latest" ]  # , "macos-latest", "windows-latest"   # enable windows after Windows driver will be added
-        python-version: [ "3.7", "3.8", "3.9", "3.10", "3.11" ]
+        python-version: [ "3.7", "3.8", "3.9", "3.10", "3.11", "3.12-dev" ]
 
     steps:
       - uses: actions/checkout@v3.5.3
@@ -140,7 +140,6 @@ jobs:
           # emulated ones
           CIBW_ARCHS_LINUX: auto aarch64
           CIBW_ARCHS_MACOS: x86_64 universal2 arm64
-          CIBW_SKIP: cp312*  # Issue #583
 
       - uses: actions/upload-artifact@v3
         with:

--- a/source/str_util.c
+++ b/source/str_util.c
@@ -458,13 +458,11 @@ offs -- offset";
 static int Py_IsWideChar(PyObject *text, Py_ssize_t offs)
 {
     const unsigned char *str;
-    Py_UNICODE *ustr;
     Py_ssize_t ret[2], str_len;
 
     if (PyUnicode_Check(text))  //text_py is unicode string
     {
-        ustr = PyUnicode_AS_UNICODE(text);
-        return (Py_GetWidth((long int)ustr[offs]) == 2);
+        return (Py_GetWidth((long int) PyUnicode_ReadChar(text, offs)) == 2);
     }
 
     if (!PyBytes_Check(text)) {
@@ -621,15 +619,13 @@ static Py_ssize_t Py_CalcWidth(PyObject *text, Py_ssize_t start_offs,
     unsigned char * str;
     Py_ssize_t i, ret[2], str_len;
     int screencols;
-    Py_UNICODE *ustr;
 
     if (PyUnicode_Check(text))  //text_py is unicode string
     {
-        ustr = PyUnicode_AS_UNICODE(text);
         screencols = 0;
  
         for(i=start_offs; i<end_offs; i++) 
-            screencols += Py_GetWidth(ustr[i]);
+            screencols += Py_GetWidth((long int) PyUnicode_ReadChar(text, i));
 
         return screencols;
     }
@@ -698,16 +694,14 @@ static int Py_CalcTextPos(PyObject *text, Py_ssize_t start_offs,
     unsigned char * str;
     Py_ssize_t i, dummy[2], str_len;
     int screencols, width;
-    Py_UNICODE *ustr;
 
     if (PyUnicode_Check(text))  //text_py is unicode string
     {
-        ustr = PyUnicode_AS_UNICODE(text);
         screencols = 0;
  
         for(i=start_offs; i<end_offs; i++)
         {
-            width = Py_GetWidth(ustr[i]);
+            width = Py_GetWidth((long int) PyUnicode_ReadChar(text, i));
             
             if (width+screencols > pref_col)
             {

--- a/source/str_util.c
+++ b/source/str_util.c
@@ -21,6 +21,8 @@
 */
 
 #define PY_SSIZE_T_CLEAN
+// Use Limited API for Python 3.7.0 release and newer
+#define Py_LIMITED_API 0x030700f0
 
 #include <Python.h>
 

